[Woo POS][App Listings] POS ASC screenshots

iamgabrielma opened the pull request:
Closes WOOMOB-1640

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
This PR adds UI tests for POS in WooCommerceScreenshots target. These screenshots will later be used by fastlane to be uploaded to ASC. We're looking into 3 screenshots:

- Dashboard view with 2 items in cart
- Totals/Payment view
- Payment Success view

## Feedback
I'd like specific feedback on 2 items:
- We didn't have any infrastructure for this screenshots, and I'm not positive that having these mocks in the WooCommerce target is the best place, but otherwise are hard to inject in order to actually mock the flows we need for the screenshots. Thoughts?
- In [commit](https://github.com/woocommerce/woocommerce-ios/pull/16355/commits/b594e990bd8b91204663b7d7b9b8ef72b8c24be6) and [commit](f614aa12c01a9ef9bd094d26ba2ebc60d0f5f6e3) I had to modify when we trigger item loading, since `onChange` only fires on value changes and not initial values, when passing mocked data into the view this wasn't being trigger as was already available from the get go. Do you see any issue with this? I haven't, but I'm not happy changing implementation for a test.

## Test Steps
- Create a `tools.fastlane` folder in your machine if you don't have one yet, in the path: `{your machine}/Library/Caches/`
- Switch the target to WooCommerceScreenshots and run the `testScreenshots()` test
- You should see tests succeeding.
- In `{your machine}/Library/Caches/tools.fastlane/screenshots` you should see the artifacts from the screenshot testing, confirm that `pos-dashboard.png`, `pos-payment.png`, and `pos-success.png` are there.

## Screenshots
| Dashboard | Payment | Success |
|--------|--------|--------|
| <img width="1180" height="820" alt="iPad (A16) - Tests-1-dark-test-pos-screenshot" src="https://github.com/user-attachments/assets/b934418b-14dc-4314-bcff-353690368d92" /> | <img width="1180" height="820" alt="iPad (A16) - Tests-2-dark-test-pos-screenshot-2" src="https://github.com/user-attachments/assets/e7fd3608-2245-4722-ad9e-cd75f6baf36d" /> | <img width="1180" height="820" alt="iPad (A16) - Tests-3-dark-test-pos-screenshot-3" src="https://github.com/user-attachments/assets/c0ac5b30-3307-4d2e-9d3c-21cecba09104" /> |

iamgabrielma committed:
fix crash on screenshot target due to unhandled action dispatch

iamgabrielma committed:
remove unused result warning

iamgabrielma committed:
Handle actions when opening POS to avoid crash (2)

iamgabrielma committed:
make compile and navigation to POS tab

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
make test POS screenshot

iamgabrielma committed:
add bypass-pos-eligibility-checks flag for UI tests

Without bypassing eligibility checks like remote FF flag or WC versions, we’ll land in the POS ineligible case and display the error view on loading the POS tab

iamgabrielma committed:
bypass POS product loading got UI test, use mocked data

iamgabrielma committed:
load mocked product image sources

iamgabrielma committed:
allow taking landscape screenshots

iamgabrielma committed:
simulate no firmware updates for ui screenshots

If we don’t handle this action in the mock, the continuation to check the firmware status never returns from `CardPresentPaymentsService.createUpdateStatePublisher`, so we’re unable to load POS fully

iamgabrielma committed:
add product to cart

iamgabrielma committed:
add multiple products

iamgabrielma committed:
checkout cart and screenshot

iamgabrielma committed:
bypass pos order syncing for screenshot

iamgabrielma committed:
fix race condition when loading products

onChange only fires on value changes, not initial values,  so when passing mocked data into the view this was only loading 10% of the times. eligibility was already .eligible and the onChange never fired

iamgabrielma committed:
waiting for payment screenshot UI test

`cardReaderViewLayout` will not pass the guard despite `posModel.orderState` and `posModel.cardReaderConnectionStatus` returning the expected values if we just pass mocked products to the order, we need to pass order items as well so the `collectCardPayment` call in `PointOfSaleAggregateModel` passes the guard that checks that order items total are not zero.

iamgabrielma committed:
process payment and make screenshot of success UI

iamgabrielma committed:
Merge branch 'task/WOOMOB-1640-post-app-listings-meta' into task/WOOMOB-1644-pos-app-listing-screenshots

# Conflicts:
#	Modules/Sources/Yosemite/Model/Mocks/ActionHandlers/MockAppSettingsActionHandler.swift

iamgabrielma committed:
inject mock item provider when needed

Determines which provider to use,  mock item provider for screenshot tests , or default PointOfSaleItemService

iamgabrielma committed:
make test pass

iamgabrielma committed:
remove comment

iamgabrielma committed:
move CardPresentPaymentServiceScreenshotMock

iamgabrielma committed:
extract mocks to their own directory

iamgabrielma committed:
log error if dependency not provided

iamgabrielma committed:
rename

iamgabrielma committed:
refactor, move oldvalue to inside onChange

we only trigger item load when onChange value changed to be .eligible, however, if eligibility was already .eligible when view appeared (like with mocked data, as is instant), onChange never fired which would make items to never load

iamgabrielma committed:
lint and periphery

iamgabrielma committed:
more periphery

for some reason is ignoring `periphery:ignore:all`

joshheald commented:
> * In [commit](https://github.com/woocommerce/woocommerce-ios/pull/16355/commits/b594e990bd8b91204663b7d7b9b8ef72b8c24be6) and [commit](f614aa12c01a9ef9bd094d26ba2ebc60d0f5f6e3) I had to modify when we trigger item loading, since `onChange` only fires on value changes and not initial values, when passing mocked data into the view this wasn't being trigger as was already available from the get go. Do you see any issue with this? I haven't, but I'm not happy changing implementation for a test.

No issues with this. It's the common pattern, and the mocked data probably just exposed a (slightly theoretical) bug.

joshheald commented on WooCommerce/Classes/POS/TabBar/POSTabCoordinator.swift:
```diff
@@ -222,11 +229,26 @@ private extension POSTabCoordinator {
+                                                      appPasswordSupportState: isAppPasswordSupported) {
+
+                let orderService: POSOrderServiceProtocol
+                if ProcessConfiguration.shouldBypassPOSOrderSyncing {
+                    orderService = POSOrderServiceScreenshotMock(currency: currencySettings.currencyCode.rawValue)
```
Yeah, I don't love this either.

Simple improvement – wrap it in an `#if debug` as well somehow?

A better (but much trickier) fix might be to set something up like the UITests use, where we mock the network rather than the service...

You could try making a factory protocol for making all this stuff, then at least we could do the ProcessConfiguration check once when creating the factory – either pass in a `POSScreenshotsDependencyFactory` or `POSDependencyFactory` – but until we have more mocked dependencies, it's probably over the top.

Best would be if we had a dependency injection framework, where we could swap these components out at will for the whole system. But Stores is the closest thing we have to that, doesn't work with async, and adding one that does is way out of scope.

Sorry, I don't have any particularly bright ideas on this one.

iamgabrielma replied:
Thanks for sharing you thoughts! I'm not sure if we'll be adding further UI tests or keep them as a one-off for these ASC screenshots. For the time being I've logged this in WOOMOB-1760

joshheald reviewed (approved):
Works as you mention... some suggestions inline

iamgabrielma committed:
periphery again

Each time we run the scan it figures out a new badge of warnings from the UI test target, this should cover them  all
